yb-voyager/cmd/exportData.go

```diff
@@ -788,14 +844,238 @@ func getInitialTableList() (map[string]string, []sqlname.NameTuple) {
+		if len(msr.TargetExportedTableListWithLeafPartitions) == 0 || msr.TargetRenameTablesMap == nil {
+			//Now add the leaf partitions to the stored table-list for the first run of export data from target
+			// as it will only have root table names and get the partitionsToRootTableMap
+			partitionsToRootTableMap, firstRunTableWithLeafsAndRoots, err = addLeafPartitionsInTableList(firstRunTableWithLeafsAndRoots, true)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to add the leaf partitions in table list: %w", err)
+			}
+		}
+	}
+	return firstRunTableWithLeafsAndRoots, partitionsToRootTableMap, nil
+}
+
+func getInitialTableList() (map[string]string, []sqlname.NameTuple, error) {
+	// store table list after filtering unsupported or unnecessary tables
+	msr, err := metaDB.GetMigrationStatusRecord()
+	if err != nil {
+		utils.ErrExit("error fetching migration status record: %v", err)
+	}
+	storedTableListNotAvailable := func() bool {
+		//if any of the list in DB is empty
+		return len(msr.TableListExportedFromSource) == 0 || len(msr.SourceExportedTableListWithLeafPartitions) == 0
+	}
+
+	if bool(startClean) || storedTableListNotAvailable() {
+		// fresh start case or the first run where we don't have a table list stored in msr
+		return fetchTablesNamesFromSourceAndFilterTableList()
+
+	}
+
+	//sunsequent run case where we will use a table-list stored in msr
+	var firstRunTableWithLeafsAndRoots []sqlname.NameTuple
 	var partitionsToRootTableMap map[string]string
-	isTableListSet := source.TableList != ""
-	partitionsToRootTableMap, finalTableList, err = addLeafPartitionsInTableList(finalTableList, isTableListSet)
+
+	firstRunTableWithLeafsAndRoots, partitionsToRootTableMap, err = retrieveFirstRunListAndPartitionsRootMap(msr)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting the first run table list and partition to root mapping: %v", err)
+	}
+
+	//guardrails around the table-list in case of re-run
+
+	//TODO: fix later this registered list also contains sequence names right now
+	//so in case some include/exclude table names are not proper we will display this registeredList as valid table names
+	registeredListObjNames, err := namereg.NameReg.GetRegisteredTableList()
+	if err != nil {
+		utils.ErrExit("error getting registered list in name registry: %v", err)
+	}
+
+	registeredList := make([]sqlname.NameTuple, 0)
+	for _, obj := range registeredListObjNames {
+		tuple, err := getNameTupleFromQualifiedObject(obj.Qualified.Quoted, obj, false)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error in getting the name tuple for the qualified object: %s: %v", obj.Qualified.Quoted, obj)
+		}
+		registeredList = append(registeredList, tuple)
+	}
+
+	rootTables := make([]sqlname.NameTuple, 0)
+	for _, v := range partitionsToRootTableMap {
+		tuple, err := namereg.NameReg.LookupTableName(v)
+		if err != nil {
+			utils.ErrExit("look up failed for the table name: %s: %v", v, err)
+		}
+		rootTables = append(rootTables, tuple)
```

Comment:
this can have duplicates right? not a problem I'm guessing?

Reply:
Yeah, in the very first implementation, I had a map but changed it to a list later. I'm not sure why I missed filtering for unique items, but it shouldn't be a problem.